banderwagon/element.go

```diff
@@ -116,49 +116,41 @@ func (p *Element) SetBytesTrusted(buf []byte) error {
 
 // computes X/Y
 func (p Element) mapToBaseField() fp.Element {
-
 	var res fp.Element
 	res.Div(&p.inner.X, &p.inner.Y)
 	return res
 }
 
-func (p Element) MapToScalarField() fr.Element {
+func (p Element) MapToScalarField(res *fr.Element) {
 	basefield := p.mapToBaseField()
 	baseFieldBytes := basefield.BytesLE()
 
-	var res fr.Element
 	res.SetBytesLE(baseFieldBytes[:])
-
-	return res
 }
 
 // Maps each point to a field element in the scalar field
-func MultiMapToScalarField(elements []*Element) []fr.Element {
+func MultiMapToScalarField(result []*fr.Element, elements []*Element) {
+	if len(result) != len(elements) {
+		panic("MultiMapToScalarField expects the result slice to be the same length of elements")
+	}
+
 	// Collect all y co-ordinates
-	var ys []fp.Element
+	ys := make([]fp.Element, len(elements))
 	for i := 0; i < int(len(elements)); i++ {
-		ys = append(ys, elements[i].inner.Y)
+		ys[i] = elements[i].inner.Y
 	}
 
 	// Invert y co-ordinates
 	yInvs := fp.BatchInvert(ys)
 
-	var scalars []fr.Element
-
 	// Multiply x by yInv
 	for i := 0; i < int(len(elements)); i++ {
 		var mappedElement fp.Element
 
 		mappedElement.Mul(&elements[i].inner.X, &yInvs[i])
 		byts := mappedElement.BytesLE()
-
-		var res fr.Element
-		res.SetBytesLE(byts[:])
-		scalars = append(scalars, res)
+		result[i].SetBytesLE(byts[:])
 	}
```

Comment:
Similar pre-allocation of slice, plus avoiding an intermediate `res` element since we can write directly in the appropriate slot of the result array.